// @vtex/api/lib/tracing/errorReporting/ErrorReport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const crypto_1 = require("crypto");
const http_1 = require("http");
const __1 = require("..");
const ErrorKinds_1 = require("./ErrorKinds");
const utils_1 = require("./utils");
class ErrorReport extends Error {
    constructor({ kind, message, originalError }) {
        var _a, _b, _c;
        super(message);
        this.kind = kind;
        this.originalError = originalError;
        this.errorId = crypto_1.randomBytes(16).toString('hex');
        this.stack = originalError.stack;
        this.message = originalError.message;
        this.errorDetails = ErrorReport.getRequestErrorMetadata(this.originalError);
        if ((_c = (_b = (_a = this.errorDetails) === null || _a === void 0 ? void 0 : _a.response) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.code) {
            this.parsedInfo = {
                serverErrorCode: this.errorDetails.response.data.code,
                ...(this.errorDetails.response.data.source ? { serverErrorSource: this.errorDetails.response.data.source } : null),
                ...(this.errorDetails.response.data.requestId ? { serverErrorRequestId: this.errorDetails.response.data.requestId } : null),
            };
        }
    }
    static create(args) {
        var _a, _b, _c;
        const kind = (_a = args.kind) !== null && _a !== void 0 ? _a : this.createGenericErrorKind(args.originalError);
        const message = (_b = args.message) !== null && _b !== void 0 ? _b : (_c = args.originalError) === null || _c === void 0 ? void 0 : _c.message;
        return new ErrorReport({
            kind,
            message,
            originalError: args.originalError,
        });
    }
    static createGenericErrorKind(error) {
        if (error.config) {
            return ErrorKinds_1.ErrorKinds.REQUEST_ERROR;
        }
        return ErrorKinds_1.ErrorKinds.GENERIC_ERROR;
    }
    static getRequestErrorMetadata(err) {
        if (!err.config) {
            return null;
        }
        const { url, method, headers: requestHeaders, params, data: requestData, timeout: requestTimeout } = err.config;
        const { status, statusText, headers: responseHeaders, data: responseData } = err.response || {};
        return {
            requestConfig: {
                data: requestData,
                headers: requestHeaders,
                method,
                params,
                timeout: requestTimeout,
                url,
            },
            response: err.response
                ? {
                    ...(responseData instanceof http_1.IncomingMessage ? { data: '[IncomingMessage]' } : { data: responseData }),
                    headers: responseHeaders,
                    status,
                    statusText,
                }
                : undefined,
        };
    }
    toObject(objectDepth = ErrorReport.DEFAULT_MAX_OBJECT_DEPTH) {
        return utils_1.truncateStringsFromObject({
            errorId: this.errorId,
            kind: this.kind,
            message: this.message,
            stack: this.stack,
            ...(this.errorDetails ? { errorDetails: this.errorDetails } : null),
            ...(this.originalError.code ? { code: this.originalError.code } : null),
        }, ErrorReport.MAX_ERROR_STRING_LENGTH, objectDepth);
    }
    injectOnSpan(span) {
        span.setTag(__1.TracingTags.ERROR, 'true');
        span.setTag(__1.TracingTags.ERROR_KIND, this.kind);
        if (this.parsedInfo) {
            span.setTag(__1.TracingTags.ERROR_SERVER_CODE, this.parsedInfo.serverErrorCode);
            if (this.parsedInfo.serverErrorSource) {
                span.setTag(__1.TracingTags.ERROR_SERVER_SOURCE, this.parsedInfo.serverErrorSource);
            }
            if (this.parsedInfo.serverErrorRequestId) {
                span.setTag(__1.TracingTags.ERROR_SERVER_REQUEST_ID, this.parsedInfo.serverErrorRequestId);
            }
        }
        span.log({ event: 'error', ...this.toObject() });
        return this;
    }
}
exports.ErrorReport = ErrorReport;
ErrorReport.MAX_ERROR_STRING_LENGTH = process.env.MAX_ERROR_STRING_LENGTH
    ? parseInt(process.env.MAX_ERROR_STRING_LENGTH, 10)
    : 8 * 1024;
ErrorReport.DEFAULT_MAX_OBJECT_DEPTH = 8;
